blockchain/blockchain_test.go

```diff
@@ -1438,3 +1441,122 @@ func TestCheckBlockChainVersion(t *testing.T) {
+	}
+	tx, err := types.NewTransactionWithMap(types.TxTypeSmartContractDeploy, values)
+	assert.Nil(t, err)
+
+	err = tx.SignWithKeys(signer, []*ecdsa.PrivateKey{key})
+	assert.Nil(t, err)
+
+	block.AddTx(tx)
+
+	contractAddr := crypto.CreateAddress(address, nonce)
+
+	// 2. Contract Execution
+	abii, err := abi.JSON(strings.NewReader(testAbiStr))
+	assert.Equal(t, nil, err)
+
+	// the contract method "sendKlay" send 1 peb to address 3 times
+	data, err := abii.Pack("sendKlay", uint32(3), address)
+	assert.Equal(t, nil, err)
+
+	values = map[types.TxValueKeyType]interface{}{
+		types.TxValueKeyNonce:    block.TxNonce(address),
+		types.TxValueKeyFrom:     address,
+		types.TxValueKeyTo:       contractAddr,
+		types.TxValueKeyAmount:   big.NewInt(0),
+		types.TxValueKeyGasLimit: gasLimit,
+		types.TxValueKeyGasPrice: gasPrice,
+		types.TxValueKeyData:     data,
+	}
+	tx, err = types.NewTransactionWithMap(types.TxTypeSmartContractExecution, values)
+	assert.Equal(t, nil, err)
+
+	err = tx.SignWithKeys(signer, []*ecdsa.PrivateKey{key})
+	assert.NoError(t, err)
+
+	block.AddTx(tx)
+}
+
+// TestCallTraceChainEventSubscription tests if the method insertChain posts a chain event correctly.
+// Scenario:
+// 1. Deploy a contract
+//   sendKlay(n uint32, receiver address): send 1 peb to `receiver` address `n` times.
+// 2. Send a smart contract execution transaction
+func TestCallTraceChainEventSubscription(t *testing.T) {
+	// configure and generate a sample block chain
+	var (
+		gendb       = database.NewMemoryDBManager()
+		key, _      = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+		address     = crypto.PubkeyToAddress(key.PublicKey)
+		funds       = big.NewInt(100000000000000000)
+		testGenesis = &Genesis{
+			Config: params.TestChainConfig,
+			Alloc:  GenesisAlloc{address: {Balance: funds}},
+		}
+		genesis = testGenesis.MustCommit(gendb)
+		signer  = types.NewEIP155Signer(testGenesis.Config.ChainID)
+	)
+	db := database.NewMemoryDBManager()
+	testGenesis.MustCommit(db)
+
+	// create new blockchain with enabled internal tx tracing option
+	blockchain, _ := NewBlockChain(db, nil, testGenesis.Config, gxhash.NewFaker(), vm.Config{Debug: true, EnableInternalTxTracing: true})
+	defer blockchain.Stop()
+
+	// subscribe a new chain event channel
+	chainEventCh := make(chan ChainEvent, 1)
+	subscription := blockchain.SubscribeChainEvent(chainEventCh)
+	defer subscription.Unsubscribe()
+
+	// generate blocks
+	blocks, _ := GenerateChain(testGenesis.Config, genesis, gxhash.NewFaker(), gendb, 1, func(i int, block *BlockGen) {
+		// Deploy a contract which can trigger internal transactions
+		genInternalTxTransaction(t, block, address, signer, key)
+	})
+
+	// insert the generated blocks into the test chain
+	if n, err := blockchain.InsertChain(blocks); err != nil {
+		t.Fatalf("failed to process block %d: %v", n, err)
+	}
+
+	// compare the published chain event with the expected test data
+	select {
```

Comment:
```suggestion

```

Reply:
I added timeout case instead of removing it.